### PR TITLE
SONARIAC-458 CMD Extend CmdTree to support exec form

### PR DESCRIPTION
**Pull request by mstachniuk**

### CODE REVIEW

File: iac-extensions/docker/src/test/java/org/sonar/iac/docker/tree/impl/CmdTreeImplTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      .notMatches("CMD \"la\", \"-bb\"]");
+  }
+
+  @Test
+  void shouldCheckParseTree() {
+    CmdTree tree = DockerTestUtils.parse("CMD [\"executable\",\"param1\",\"param2\"]", DockerLexicalGrammar.CMD);
+    assertThat(tree.getKind()).isEqualTo(DockerTree.Kind.CMD);
+    assertThat(tree.keyword().value()).isEqualTo("CMD");
+
+    assertThat(tree.cmdArguments().stream().map(TextTree::value)).containsExactly("\"executable\"", "\"param1\"", "\"param2\"");
```
Comment (mstachniuk) [inline iac-extensions/docker/src/test/java/org/sonar/iac/docker/tree/impl/CmdTreeImplTest.java:65]: Do you think, is it good to have here arguments surrounded by double quotes?

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/DockerTree.java [hunk 2]
```diff
@@ -39,6 +39,9 @@ enum Kind implements GrammarRuleKey {
     LABEL(LabelTree.class),
     ENV(EnvTree.class),
     KEY_VALUE_PAIR(KeyValuePairTree.class),
+    CMD(CmdTree.class),
+    EXEC_FORM(ExecFormTree.class),
```
Comment (mstachniuk) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/DockerTree.java:43]: This one is unused, is it good?

Comment (mstachniuk) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/DockerTree.java:43]: Nevermind, additional tests covered it

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/TreeFactory.java [hunk 3]
```diff
@@ -20,35 +20,45 @@
...
 import org.sonar.iac.docker.tree.impl.LabelTreeImpl;
 import org.sonar.iac.docker.tree.impl.MaintainerTreeImpl;
+import org.sonar.iac.docker.tree.impl.SeparatedListImpl;
 import org.sonar.iac.docker.tree.impl.StopSignalTreeImpl;
 import org.sonar.iac.docker.tree.impl.WorkdirTreeImpl;
 import org.sonar.iac.docker.tree.impl.PortTreeImpl;
 
 public class TreeFactory {
 
+  @SuppressWarnings("java:S1172")
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/TreeFactory.java:61]: Maybe add a comment about what this suppressed rule is.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/grammar/DockerGrammar.java [hunk 4]
```diff
@@ -189,4 +192,38 @@ public KeyValuePairTree KEY_VALUE_PAIR_WITH_EQUALS() {
       f.keyValuePairEquals(b.token(STRING_LITERAL), b.token(Punctuator.EQU), b.token(STRING_LITERAL))
     );
   }
+
+  public CmdTree CMD() {
+    return b.<CmdTree>nonterminal(DockerLexicalGrammar.CMD).is(
+      f.cmd(
+        b.token(DockerKeyword.CMD),
+        b.optional(EXEC_FORM())
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/grammar/DockerGrammar.java:200]: I'm supprised that the actual command is optional, but it seems valid.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/TreeFactory.java [hunk 5]
```diff
@@ -104,4 +114,56 @@ public KeyValuePairTree keyValuePair(SyntaxToken key, SyntaxToken value) {
...
+            separators.add(comaAndLiteral.first());
+            elements.add(new ExecFormLiteralTreeImpl(comaAndLiteral.second()));
+          }
+        }
+    }
+
+    return new ExecFormTreeImpl(leftBracket, separatedList, rightBracket);
+  }
+
+  public <T, U> Tuple<T, U> newTuple(T first, U second) {
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/TreeFactory.java:145]: Maybe tuple() is enough. We don't use the new keyword somewhere else.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/CmdTree.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.docker.tree.api;
+
+
+import java.util.List;
+
+public interface CmdTree extends InstructionTree {
+
+  ExecFormTree execFormTree();
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/CmdTree.java:27]: execForm()?

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/ExecFormTree.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.docker.tree.api;
+
+public interface ExecFormTree extends DockerTree {
+  SyntaxToken leftBracket();
+  SeparatedList<ExecFormLiteralTree> literals();
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/ExecFormTree.java:24]: literals seems quite generic for this class. However, I don't have a better name for now.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/impl/CmdTreeImpl.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+import org.sonar.iac.common.api.tree.Tree;
+import org.sonar.iac.docker.tree.api.CmdTree;
+import org.sonar.iac.docker.tree.api.ExecFormLiteralTree;
+import org.sonar.iac.docker.tree.api.ExecFormTree;
+import org.sonar.iac.docker.tree.api.SeparatedList;
+import org.sonar.iac.docker.tree.api.SyntaxToken;
+
+public class CmdTreeImpl extends InstructionTreeImpl implements CmdTree {
+
+  private final ExecFormTree execFormTree;
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/impl/CmdTreeImpl.java:34]: Same as above: IMO execForm is enough

File: iac-extensions/docker/src/test/java/org/sonar/iac/docker/parser/grammar/DockerLexicalGrammarTest.java [hunk 9]
```diff
@@ -67,4 +67,17 @@ void testStringEol() {
...
+    Assertions.assertThat(DockerLexicalGrammar.EXEC_FORM)
+      .matches("[]")
+      .matches("[\"foo\"]")
+      .matches("[\"foo\", \"bar\"]")
+      .matches("[\"foo\",\"bar\"]")
+      .matches("[\"foo\" , \"bar\"]")
+      .notMatches("[\"foo\" \"bar\"]")
+      .notMatches("[\"foo\", \"bar\",]")
+      .notMatches("");
+  }
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/test/java/org/sonar/iac/docker/parser/grammar/DockerLexicalGrammarTest.java:82]: Do we have to do this test? Is ExecFormTreeImplTest not enough?

File: iac-extensions/docker/src/test/java/org/sonar/iac/docker/tree/impl/CmdTreeImplTest.java (hunk 1 above)
Comment (nils-werner-sonarsource) [reply iac-extensions/docker/src/test/java/org/sonar/iac/docker/tree/impl/CmdTreeImplTest.java:62]: I think, it's okay for now.

**Review by nils-werner-sonarsource — APPROVED** [review]
LGTM! Please see my comments and questions.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/grammar/DockerGrammar.java (hunk 4 above)
Comment (mstachniuk) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/grammar/DockerGrammar.java:200]: Yes, docker for `CMD` produce:
`CMD ["/bin/sh" "-c" ""]`

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/ExecFormTree.java (hunk 7 above)
Comment (mstachniuk) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/ExecFormTree.java:24]: mee too, let's keep it for now

File: iac-extensions/docker/src/test/java/org/sonar/iac/docker/parser/grammar/DockerLexicalGrammarTest.java (hunk 9 above)
Comment (mstachniuk) [reply iac-extensions/docker/src/test/java/org/sonar/iac/docker/parser/grammar/DockerLexicalGrammarTest.java:82]: good catch, I will merge those

**Pull request merged**